lib/srv/db/spanner/engine.go

```diff
@@ -0,0 +1,185 @@
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package spanner
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"cloud.google.com/go/spanner/apiv1/spannerpb"
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+	gtransport "google.golang.org/api/transport/grpc"
+	"google.golang.org/grpc"
+
+	"github.com/gravitational/teleport/api/utils/grpc/interceptors"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// NewEngine creates a new Spanner engine.
+func NewEngine(ec common.EngineConfig) common.Engine {
+	return &Engine{
+		EngineConfig: ec,
+	}
+}
+
+// Engine implements common.Engine.
+type Engine struct {
+	// EngineConfig is the common database engine configuration.
+	common.EngineConfig
+	// clientConn is a client connection.
+	clientConn net.Conn
+	// clientConnContext is the connection context passed into HandleConnection.
+	// It's used to control the upstream connection (and credentials refresher)
+	// lifetime, and cancelled when the engine exits.
+	clientConnContext context.Context
+	// sessionCtx is current session context.
+	sessionCtx *common.Session
+	// instanceID is used for authz and is of the form
+	// projects/<project-id>/instances/<instance-id>/,
+	// where the project ID and instance ID come from the Teleport database's
+	// GCP metadata.
+	// Every RPC that specifies a database name must be prefixed by this or the
+	// RPC will be denied.
+	instanceID string
+	// server is the underlying gRPC server.
+	server *grpc.Server
+
+	// mu guards access to gcloudClient and, until the session starts,
+	// sessionStartErr.
+	mu sync.Mutex
+	// gcloudConn is an upstream client to Google cloud API used to forward RPCs
+	// to the real Spanner API.
+	gcloudClient gtransport.ConnPool
+	// sessionStartErr is the error (or nil) from the first RPC access check
+	// that starts the session.
+	sessionStartErr error
+
+	// sessionStarted is true when the session start event has been emitted.
+	sessionStarted atomic.Bool
+	// closeOnce is used to early close the gRPC only once.
+	closeOnce sync.Once
+
+	// UnimplementedSpannerServer is embedded for forward interface compat.
+	spannerpb.UnimplementedSpannerServer
```

Comment:
It means users will get a not implemented error message, which would be accurate... but also means we dont find out about a new method until people ask us to implement it. That could be a good or bad thing depending on your perspective ;)

I am hesitant to block our CI if they ever add methods though, we should be able to bump our spanner dependency without any code changes.

> Perhaps we can test is somehow, to make the break scoped to that single test rather than whole package?

I'm not sure how to test that without a type assertion, which breaks test compilation... I don't have a good answer tbh. I'll have to think on this some more.